### PR TITLE
[ISSUE #1120]Add new feature: support namespace

### PR DESCRIPTION
**Pull request by lollipopjin**
## What is the purpose of the change
Implement: [Issue #1120 ](https://github.com/apache/rocketmq/issues/1120)

## Brief changelog
Add new feature: support namespace

## Verifying this change

Follow this checklist to help us incorporate your contribution quickly and easily. Notice, `it would be helpful if you could finish the following 5 checklist(the last one is not necessary)before request the community to review your PR`.

- [x] Make sure there is a [Github issue](https://github.com/apache/rocketmq/issues) filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue. 
- [x] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test(over 80% coverage) to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/rocketmq/tree/master/test).
- [x] Run `mvn -B clean apache-rat:check findbugs:findbugs checkstyle:checkstyle` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.
- [x] If this contribution is large, please file an [Apache Individual Contributor License Agreement](http://www.apache.org/licenses/#clas).

### CODE REVIEW

File: common/pom.xml [hunk 1]
```diff
@@ -37,5 +37,9 @@
             <groupId>${project.groupId}</groupId>
             <artifactId>rocketmq-remoting</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.apache.commons</groupId>
+            <artifactId>commons-lang3</artifactId>
```
Comment (vongosling) [inline common/pom.xml]: Be sure need to import lang3  here?

**Comment by vongosling** [comment]
Could we consider minimize the pollution if we introduce another new field in the context?

File: common/src/main/java/org/apache/rocketmq/common/protocol/NamespaceUtil.java [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+    public static String getNamespaceFromRequest(RemotingCommand request) {
+        if (null == request || null == request.getExtFields()) {
+            return null;
+        }
+
+        String namespace;
+
+        switch (request.getCode()) {
+            case RequestCode.SEND_MESSAGE_V2:
+                namespace = request.getExtFields().get("n");
```
Comment (zongtanghu) [inline common/src/main/java/org/apache/rocketmq/common/protocol/NamespaceUtil.java]: Here,will you add a field called "n" in the SendMessageRequestHeaderV2 Class File? @lollipopjin

File: common/src/main/java/org/apache/rocketmq/common/protocol/NamespaceUtil.java [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+                break;
+            default:
+                namespace = request.getExtFields().get("namespace");
```
Comment (zongtanghu) [inline common/src/main/java/org/apache/rocketmq/common/protocol/NamespaceUtil.java]: And,where do you add the field called"namesapce" into the request's extFields? @lollipopjin

File: common/src/main/java/org/apache/rocketmq/common/protocol/NamespaceUtil.java (hunk 2 above)
Comment (lollipopjin) [reply common/src/main/java/org/apache/rocketmq/common/protocol/NamespaceUtil.java]: Hi，I remove the  redundant parts in NamespaceUtil.java and do some refactors in the new commit.

File: common/src/main/java/org/apache/rocketmq/common/protocol/NamespaceUtil.java (hunk 3 above)
Comment (lollipopjin) [reply common/src/main/java/org/apache/rocketmq/common/protocol/NamespaceUtil.java]: The redundant parts have been removed.

File: common/src/main/java/org/apache/rocketmq/common/protocol/NamespaceUtil.java (hunk 2 above)
Comment (zongtanghu) [reply common/src/main/java/org/apache/rocketmq/common/protocol/NamespaceUtil.java]: Okay,it's much more better!I will review it later.Thank you!

File: common/src/main/java/org/apache/rocketmq/common/protocol/NamespaceUtil.java (hunk 3 above)
Comment (zongtanghu) [reply common/src/main/java/org/apache/rocketmq/common/protocol/NamespaceUtil.java]: Okay,it's much more better!

**Comment by zongtanghu** [comment]
LGTM

**Comment by duhenglucky** [comment]
LGTM

**Comment by vongosling** [comment]
Very fragmented implementation for multi-tenant case, we could merge it as the workaroud. Optimize here in RocketMQ 5.0.

File: client/src/main/java/org/apache/rocketmq/client/consumer/DefaultMQPullConsumer.java [hunk 4]
```diff
@@ -250,6 +265,7 @@ public void setRegisterTopics(Set<String> registerTopics) {
     @Override
     public void sendMessageBack(MessageExt msg, int delayLevel)
         throws RemotingException, MQBrokerException, InterruptedException, MQClientException {
+        msg.setTopic(withNamespace(msg.getTopic()));
```
Comment (mekswhy) [inline client/src/main/java/org/apache/rocketmq/client/consumer/DefaultMQPullConsumer.java]: Check msg is null and throw MQClientException if necessary.

Comment (duhenglucky) [reply client/src/main/java/org/apache/rocketmq/client/consumer/DefaultMQPullConsumer.java]: yes, this issue has been resolved,  @ShannonDing added the check logic before send message

**Pull request merged**